### PR TITLE
Reimplement Result object with integrated visualization

### PR DESCRIPTION
**Pull request by vuolleko**
Rejection sampling and SMC now return a Result object as in earlier version. Integrates basic visualization.

### CODE REVIEW

File: elfi/methods/methods.py [hunk 1]
```diff
@@ -478,18 +476,28 @@ def update(self, batch, batch_index):
...
         if self.state['samples'] is None:
             raise ValueError('Nothing to extract')
 
         # Take out the correct number of samples
-        samples = dict()
-        for k, v in self.state['samples'].items():
-            samples[k] = v[:self.objective['n_samples']]
+        n_samples = self.objective['n_samples']
+        samples = [self.state['samples'][p][:n_samples]
+                   for p in self.outputs]
```
Comment (lintusj1) [inline elfi/methods/methods.py]: Wondering the change here. This potentially changes the behaviour, where the earlier took all the keys from `samples`, and this only `self.outputs`.

File: elfi/results/result.py [hunk 2]
```diff
@@ -19,14 +19,25 @@ class Result(object):
...
+            else:
+                self.samples[n] = samples_list[ii]
+        if distance is not None:
+            self.distance = distance
+        # self.n_samples = len(list(self.samples.values())[0])
+        self.n_samples = len(samples_list[0])
+        self.n_params = len(samples_list)
 
         # get name of the ABC method
         stack10 = inspect.stack()[1][0]
```
Comment (lintusj1) [inline elfi/results/result.py]: Would it be cleaner to pass the method name to the `__init__`?

File: elfi/visualization.py [hunk 3]
```diff
@@ -1,3 +1,7 @@
+import numpy as np
+import matplotlib.pyplot as plt
+from collections import OrderedDict
+
 from elfi.model.elfi_model import ElfiModel, NodeReference
```
Comment (lintusj1) [inline elfi/visualization.py]: Could this file be combined with `plotting.py` so that we don't have two different files for plotting.

Comment (vuolleko) [inline elfi/visualization.py]: Yes, definitely. But maybe better to leave that until we have a better picture of what is needed.

File: elfi/methods/methods.py [hunk 4]
```diff
@@ -622,21 +640,22 @@ def _new_round(self):
                                        threshold=self.current_population_threshold)
 
     def _extract_population(self):
-        pop = self._rejection.extract_result()
+        pop = self._rejection.extract_result("Rejection within SMC-ABC")
```
Comment (lintusj1) [inline elfi/methods/methods.py]: I would do the population renaming here.

Comment (lintusj1) [inline elfi/methods/methods.py]: `pop.method = ...`

**Pull request merged**